blockchain/validate.go

```diff
@@ -774,14 +774,34 @@ func (b *BlockChain) checkBlockContext(block *btcutil.Block, prevNode *blockNode
 
 	fastAdd := flags&BFFastAdd == BFFastAdd
 	if !fastAdd {
+		medianTime := b.BestSnapshot().MedianTime
+
+		// Obtain the latest state of the deployed CSV soft-fork in
+		// order to properly guard the new validation behavior based on
+		// the current BIP 9 version bits state.
+		csvState, err := b.ThresholdState(chaincfg.DeploymentCSV)
+		if err != nil {
+			return err
+		}
+
+		// Once the CSV soft-fork is fully active, we'll switch to
+		// using the current median time past of the past block's
+		// timestamps for all lock-time based checks.
+		var blockTime time.Time
+		if csvState == ThresholdActive {
+			blockTime = medianTime
+		} else {
```

Comment:
nit: Perhaps it's just a matter of taste, but I find the set and override more readable since it's shorter:

```Go
blockTime := header.Timestamp
if csvState == ThresholdActive {
	blockTime = medianTime
}
```